index.js

```diff
@@ -9,41 +9,96 @@ function defaultInspect() {
+    // perform some stack cleanup
+    this.stack = this.stack.replace(/^.+?NewError(((?!=NewError).+)[\r\n])/gm, '').replace(new RegExp('.+?' + this.name + '.+?[\r\n]+'), '')
+  }
 }
 
-function ErrorMaker(name, ParentError) {
-  function NewError(message) {
-    if (!(this instanceof NewError))
-      return new NewError(message)
-
-    // Use a try/catch block to capture the stack trace. Capturing the stack trace here is
-    // necessary, otherwise we will get the stack trace at the time the new error class was created,
-    // rather than when it is instantiated.  We add `message` and `name` so that the stack trace
-    // string will match our current error class.
+function BaseErrorConstructor (message) {
+  // Use a try/catch block to capture the stack trace. Capturing the stack trace here is
+  // necessary, otherwise we will get the stack trace at the time the new error class was created,
+  // rather than when it is instantiated.  We add `message` and `name` so that the stack trace
+  // string will match our current error class.
+  if (Error.captureStackTrace) {
+      Error.captureStackTrace(this);
+  } else {
     try {
       throw new Error(message)
+    } catch (err) {
+      this.stack = err.stack;
     }
-    catch (err) {
-      err.name = name
-      this.stack = err.stack
-    }
-
+  }
     this.message = message || ''
-    this.name = name
 
     this.reformatStack();
+}
+
+function ErrorMaker(name, ParentError, ctr) {
+  function NewError(error, message) {
+    if (!(this instanceof NewError)) {
+      return new NewError(error, message)
+    }
+
+    var args = Array.prototype.slice.call(arguments);
+
+    // hard processing of nested error
+    if (error instanceof Error) {
+      this.nested = error;
+      // Error does not provide error serializer and this makes error logging useless
+      if (!(error.toJSON)) {
+        // provide custom serialization
+        Object.defineProperty(error, 'toJSON', {
+            value: defaultToJson,
+            configurable: true,
+            writable: true
+        });
+      }
+      // do not pass error down the parent hierarchy
+      // it will be available as property for other constructors
+      args.shift();
+    }
+
+    if (ParentError) {
+      // invoke parent constructors before the current one
+      ParentError.apply(this, args);
+    }
+    if (ctr) {
+      // final invocation
+      ctr.apply(this, args);
+    }
   }
 
-  NewError.prototype = new (ParentError || Error)()
-  NewError.prototype.constructor = NewError
+  // add better looking stack name, instead of 'NewError'
+  Object.defineProperty(NewError, "name", {value: name});
+
+  NewError.prototype = new (ParentError || Error)();
+  NewError.prototype.constructor = NewError;
   NewError.prototype.inspect = defaultInspect;
-  NewError.prototype.name = name
+  NewError.prototype.name = name;
   NewError.prototype.reformatStack = reformatStack;
+  NewError.prototype.toJSON = defaultToJson;
```

Comment:
I think `toJSON` should be added by the end user. I'd prefer to just keep this as a barebones library that doesn't extend functionality of Error in any way expect for making it extendable.